File: app/formatters.py
"""
Formatters translate a raw payload dict into the specific JSON shape each
destination type expects. Each formatter returns a plain dict that gets
JSON-serialised into the outgoing request body.
 
Adding a new destination type means adding one function here and one entry
in FORMATTERS, nothing else in the codebase changes.
"""

from __future__ import annotations

from typing import Any

def _truncate(text: str, limit: int) -> str:
    """Discord/Slack have field limits, so truncate if needed"""
    return text if len(text) <= limit else text[:limit - 1] + "..."

def format_discord(payload: dict[str, Any], event_type: str | None) -> dict[str, Any]:
    """
    Discord payload formatter using embeds
    https://discord.com/developers/docs/resources/channel#embed-object
    """

    title = event_type or "Webhook Event"
    description = payload.get("message") or payload.get("text") or payload.get("description") or ""

    fields = [
        {
            "name": _truncate(str(k), 256),
            "value": _truncate(str(v), 1024),
            "inline": True,
        }
        for k, v in payload.items()
        if k not in ("message", "text", "description")
    ]

    embed: dict[str, Any] = {
        "title": _truncate(title, 256),
        "color": 5793266,  # a neutral blue
        "fields": fields[:25],  # Discord hard limit
    }

    if description:
        embed["description"] = _truncate(description, 4096)

    return {"embeds", [embed]}

def format_slack(payload: dict[str, Any], event_type: str | None) -> dict[str, Any]:
    """
    Slack payload formatter using blocks
    https://api.slack.com/block-kit
    """

    blocks: list[dict[str, Any]] = []

    if event_type:
        blocks.append({
            "type": "header",
            "text": {
                "type": "plain_text",
                "text": _truncate(event_type, 150),
            },
        })

    for key, value in payload.items():
        text = f"*{key}*\n{_truncate(str(value), 2900)}"
        blocks.append({
            "type": "section",
            "text": {
                "type": "mrkdwn",
                "text": text,
            },
        })

    
    # Slack recommends a divider at the end for readability
    if blocks:
        blocks.append({"type": "divider"})

    return {"blocks": blocks[:50]} # Slack hard limit

def format_generic(payload: dict[str, Any], event_type: str | None) -> dict[str, Any]:
    """
    Passes the payload through as-is, adding "event_type" field if provided.
    """

    body = dict(payload)
    if event_type:
        body["event_type"] = event_type
    return body

FORMATTERS = {
    "discord": format_discord,
    "slack": format_slack,
    "generic": format_generic,
}

def format_payload(destination_type: str, payload: dict[str, Any], event_type: str | None) -> dict[str, Any]:
    formatter = FORMATTERS.get(destination_type, format_generic)
    return formatter(payload, event_type)

File: internal/formatters/formatters.go
// Formatters translate a raw payload into a specific JSON shape for a given destination.
//
// Python used a plain dict of functions:
//	FORMATTERS = {"discord": format_discord, ...}
//
// Go expresses the same idea as a map of typed function values:
//	var formatters = map[models.DestinationType]FormatterFunc{...}

package formatters

import (
	"fmt"
	"webhook-dispatcher/internal/models"
)

type FormatterFunc func(payload map[string]any, eventType *string) map[string]any

var registry = map[models.DestinationType]FormatterFunc{}

// Route to the correct formatter, falling back to a generic formatter.
func Format(destType models.DestinationType, payload map[string]any, eventType *string) map[string]any {
	if fn, ok := registry[destType]; ok {
		return fn(payload, eventType)
	}
	return FormatGeneric(payload, eventType)
}

func truncate(s string, limit int) string {
	runes := []rune(s)
	if len(runes) <= limit {
		return s
	}
	return string(runes[:limit-1]) + "…"
}

func strVal(v any) string {
	return fmt.Sprintf("%v", v)
}

func FormatGeneric(payload map[string]any, eventType *string) map[string]any {
	out := make(map[string]any, len(payload)+1)
	for k, v := range payload {
		out[k] = v
	}
	if eventType != nil && *eventType != "" {
		out["event_type"] = *eventType
	}
	return out
}

func FormatDiscord(payload map[string]any, eventType *string) map[string]any {
	title := "Webhook Event"
	if eventType != nil && *eventType != "" {
		title = *eventType
	}

	description := ""
	for _, key := range []string{"message", "text", "description"} {
		if v, ok := payload[key]; ok {
			description = strVal(v)
			break
		}
	}

	skipKeys := map[string]bool{"message": true, "text": true, "description": true}
	var fields []map[string]any
	for k, v := range payload {
		if skipKeys[k] || len(fields) >= 25 {
			break
		}
		fields = append(fields, map[string]any{
			"name":  truncate(k, 256),
			"value": truncate(strVal(v), 1024),
			"inline": true,
		})
	}

	embed := map[string]any{
		"title": truncate(title, 256),
		"color":  5793266,
		"fields": fields,
	}
	if description != "" {
		embed["description"] = truncate(description, 4096)
	}

	return map[string]any{"embeds": []any{embed}}
}

func FormatSlack(payload map[string]any, eventType *string) map[string]any {
	var blocks []any

	if eventType != nil && *eventType != "" {
		blocks = append(blocks, map[string]any{
			"type": "header",
			"text": map[string]any{
				"type": "plain_text",
				"text": truncate(*eventType, 150),
			},
		})
	}

	for k, v := range payload {
		if len(blocks) >= 49 {
			break
		}
		text := fmt.Sprintf("*%s*\n%s", k, truncate(strVal(v), 2900))
		blocks = append(blocks, map[string]any{
			"type": "section",
			"text": map[string]any{"type": "mrkdwn", "text": text},
		})
	}

	if len(blocks) > 0 {
		blocks = append(blocks, map[string]any{"type": "divider"})
	}

	return map[string]any{"blocks": blocks}
}